### PR TITLE
Sharded Accelerator 1/n: Expose clip gradients to plugins via abstract class

### PR DESCRIPTION
**Pull request by SeanNaren**
Ties to #4178 

## What does this PR do?

To ensure we have the appropriate behaviour in clip gradients, I've made it a responsibility of the plugin if it exists in the accelerator. I've also simplified the logic to reflect what should happen in AMP/Apex cases using the torch built in `clip_grad_norm_` function.

This is required because sharded accelerator will need a custom clip gradient function :)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SeanNaren** [comment]
Need to sort out APEX fp16 (with level set to O2 all parameters are in FP16, I need to change the epsilon to match this)

**Review by tchaton — CHANGES_REQUESTED** [review]
Great ! Just one question about clipping for cpu.

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -149,17 +143,19 @@ def _clip_gradients(self, optimizer, clip_val=None):
...
             return
 
-        model = self.trainer.get_model()
-        if self.trainer.amp_backend == AMPType.APEX:
-            parameters = amp.master_params(optimizer)
+        if self.trainer.amp_backend:
+            self.trainer.precision_connector.backend.clip_gradients(grad_clip_val, optimizer)
         else:
-            parameters = model.parameters()
+            self._clip_gradients_with_tpu_support(grad_clip_val)
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:152]: This was the default behaviour, with a TODO left from a previous lifetime I guess 😅 (you can see the TODO above in `clip_gradients`) it does the same thing as normal clipping, but with operations that are supported on the TPU.

I think if we're alright with the change being in this PR, we can default to `torch.nn.utils.clip_grad_norm_` here, and then override in the TPU class this additional logic

File: pytorch_lightning/plugins/native_amp.py [hunk 2]
```diff
@@ -51,3 +53,9 @@ def training_step(self, fx, args):
         with torch.cuda.amp.autocast():
             output = fx(*args)
         return output
+
+    def clip_gradients(self, grad_clip_val, optimizer):
+        model = self.trainer.get_model()
+        max_norm = grad_clip_val
+        norm_type = float(2.0)
```
Comment (SkafteNicki) [inline pytorch_lightning/plugins/native_amp.py:60]: `2.0` is already a float?

File: pytorch_lightning/plugins/native_amp.py [hunk 3]
```diff
@@ -51,3 +53,9 @@ def training_step(self, fx, args):
...
+        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=max_norm, norm_type=norm_type)
```
Comment (SkafteNicki) [inline pytorch_lightning/plugins/native_amp.py:61]:
```suggestion
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=grad_clip_val, norm_type=2.0)
```

File: pytorch_lightning/plugins/precision_plugin.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import abc
```
Comment (SkafteNicki) [inline pytorch_lightning/plugins/precision_plugin.py:1]: Remember to add our copyright text

**Review by SkafteNicki — APPROVED** [review]
Smaller comments, LGTM

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py]: I made the change, please review @tchaton!

File: pytorch_lightning/accelerators/tpu_accelerator.py [hunk 5]
```diff
@@ -261,10 +264,31 @@ def optimizer_step(self, optimizer, batch_idx, opt_idx, lambda_closure):
             using_lbfgs=is_lbfgs
         )
 
-    def clip_gradients(self, optimizer, clip_val=None):
-        # apply clip gradients
-        # TODO: separate TPU case from here
-        self._clip_gradients(optimizer, clip_val)
+    def _clip_gradients(self, optimizer, grad_clip_val):
```
Comment (tchaton) [inline pytorch_lightning/accelerators/tpu_accelerator.py:267]: Can we have only one which takes `EPSILON`, either `EPSILON` or `FP16_EPSILON `. Less duplicated code.

**Review by tchaton — CHANGES_REQUESTED** [review]

File: pytorch_lightning/accelerators/tpu_accelerator.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/tpu_accelerator.py:267]: Maybe makes gradient_clipping_epsilon a property.

Comment (SeanNaren) [reply pytorch_lightning/accelerators/tpu_accelerator.py:265]: Moved to properties contained in the respective classes

Comment (SeanNaren) [reply pytorch_lightning/accelerators/tpu_accelerator.py:265]: The functions are different, the one in Apex is true to the torch code and the tpu one is based on the TPU guide suggestions, combining them is tricky and I think the logic should be kept separate...

**Review by tchaton — APPROVED** [review]
Great PR !

File: pytorch_lightning/accelerators/tpu_accelerator.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/accelerators/tpu_accelerator.py:265]: We have the epsilons defined already

File: pytorch_lightning/accelerators/accelerator.py [hunk 6]
```diff
@@ -139,48 +130,22 @@ def optimizer_zero_grad(self, batch_idx, optimizer, opt_idx):
...
-        # this code is a modification of torch.nn.utils.clip_grad_norm_
-        # with TPU support based on https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md
         if grad_clip_val <= 0:
             return
+        self._clip_gradients(optimizer, grad_clip_val)
 
-        model = self.trainer.get_model()
-        if self.trainer.amp_backend == AMPType.APEX:
-            parameters = amp.master_params(optimizer)
+    def _clip_gradients(self, optimizer, grad_clip_val):
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:143]: can we add typing?

File: pytorch_lightning/accelerators/tpu_accelerator.py [hunk 7]
```diff
@@ -261,10 +262,31 @@ def optimizer_step(self, optimizer, batch_idx, opt_idx, lambda_closure):
...
+        # this code is a modification of torch.nn.utils.clip_grad_norm_
+        # with TPU support based on https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md
+        model = self.trainer.get_model()
+        parameters = model.parameters()
+        max_norm = grad_clip_val
+        norm_type = 2.0
```
Comment (Borda) [inline pytorch_lightning/accelerators/tpu_accelerator.py:271]: shall it be an argument?

File: pytorch_lightning/accelerators/tpu_accelerator.py [hunk 8]
```diff
@@ -261,10 +262,31 @@ def optimizer_step(self, optimizer, batch_idx, opt_idx, lambda_closure):
...
+
+        if isinstance(parameters, torch.Tensor):
+            parameters = [parameters]
+        parameters = list(filter(lambda p: p.grad is not None, parameters))
+
+        if norm_type == math.inf:
```
Comment (Borda) [inline pytorch_lightning/accelerators/tpu_accelerator.py:277]: how this can happen with `norm_type = 2.0` above?

File: pytorch_lightning/plugins/apex.py [hunk 9]
```diff
@@ -98,3 +101,38 @@ def configure_apex(self, amp, model, optimizers, amp_level):
         """
         model, optimizers = amp.initialize(model, optimizers, opt_level=amp_level)
         return model, optimizers
+
+    def clip_gradients(self, grad_clip_val, optimizer):
```
Comment (Borda) [inline pytorch_lightning/plugins/apex.py:105]:
```suggestion
    def _clip_gradients(self, grad_clip_val, optimizer):
```
it shall be parent overwrite here, right?

File: pytorch_lightning/accelerators/accelerator.py [hunk 10]
```diff
@@ -139,48 +130,22 @@ def optimizer_zero_grad(self, batch_idx, optimizer, opt_idx):
...
-                torch.norm(p.grad.data.to(device), norm_type, out=out[i])
-            total_norm = torch.norm(out, norm_type)
-
-        eps = EPSILON_FP16 if self.trainer.precision == 16 else EPSILON
-        clip_coef = torch.tensor(max_norm, device=device) / (total_norm + eps)
-        clip_coef = torch.min(clip_coef, torch.ones_like(clip_coef))
-        for p in parameters:
-            p.grad.data.mul_(clip_coef.to(p.grad.data.device))
+            model = self.trainer.get_model()
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=grad_clip_val, norm_type=2.0)
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:148]: I wonder if it would make sense to have a precision plugin for 32bit for the default case. 
Then we wouldn't need this if/else block and could just call 
`self.trainer.precision_connector.backend.clip_gradients(grad_clip_val, optimizer)`
just as a refactoring idea for the future.

**Review by awaelchli — APPROVED** [review]

**Review by Borda — COMMENTED** [review]
I feel a bit confused about what is going to be overwritten for Accelerator and plugin, one seems to overwrite in child `clip_gradients` the other `_ clip_gradients`

File: pytorch_lightning/accelerators/accelerator.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:148]: (talking about this block, github selected more)
```
def _clip_gradients(self, optimizer, grad_clip_val):
        if self.trainer.amp_backend:
            self.trainer.precision_connector.backend.clip_gradients(grad_clip_val, optimizer)
        else:
```

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:148]: I agree fully, a precision 32 plugin makes the most sense because the accelerators shouldn't need to care about these details (It goes into @justusschock's refactor proposal of making sure the accelerators are hardware only).

Out of the scope of this PR, but definitely something low hanging to reduce liability of issues in the accelerators

File: pytorch_lightning/accelerators/tpu_accelerator.py (hunk 8 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/tpu_accelerator.py:277]: This was the default code from before and I agree, it just adds confusion. I'll remove this from the code entirely!

File: pytorch_lightning/plugins/apex.py (hunk 9 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/apex.py:105]: This doesn't inherit from the accelerator class, but is called within the plugin. Let's do some renaming since there is a lot of confusion here.

File: pytorch_lightning/accelerators/accelerator.py (hunk 10 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:148]: I'm going to do this in my refactor proposal :)

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/apex.py (hunk 9 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/apex.py]: After thinking through I don't think it's an easy rename any suggestions @Borda? otherwise I'll merge and revise in a separate PR when needed.

Comment (Borda) [reply pytorch_lightning/plugins/apex.py]: not sure, maybe add some docstring with an explanation...
